File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode reversedHead = new ListNode();
        ListNode temp = null;

        while(head != null){
            
            temp = new ListNode();
            temp.val = head.val;

            if(reversedHead.next == null)
                reversedHead.next = temp;
            else{
                temp.next = reversedHead.next;
                reversedHead.next = temp;
            }

            head = head.next;
        }

        return reversedHead.next;
    }
}

File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        def recursive(root, past):
            if not root:
                return past

            temp = root.next
            root.next = past

            return recursive(temp, root)
        
        return recursive(head, None)